### PR TITLE
[Perf][GLM-5.3-Flash] Decode hot-path cleanups: strided KDA recurrent inputs, NoPE MQA query without concat, no duplicate router GEMM

### PR DESCRIPTION
**Pull request by JaredforReal**
## Purpose

One of three independent GLM-5.3-Flash perf PRs 

Three small, independent decode hot-path cleanups found while profiling GLM-5.3-Flash (`zai-org/GLM-5.3-Flash`, FP8, TP4 on 4x GB300, `FLASHINFER_MLA_SPARSE`):

1. **KDA decode: read token-strided q/k/v/beta in `fused_recurrent_gated_delta_rule_fwd_kernel`.** The decode path hands the recurrent kernel column slices of the merged `q|k|v` conv output and of the fused `qkvbfg_a` projection (beta); `fused_recurrent_kda` made each contiguous, i.e. 4 copy kernels per KDA layer per step (34 layers; ~11 us/layer at 64 requests, ~0.4 ms/step). The Triton kernel now takes explicit token strides; contiguous callers are unchanged. The stride check is pure integer arithmetic (no views) because the KDA forward is a CUDA-graph break and runs eagerly.
2. **MLA: write the absorbed MQA query token-major and skip the empty RoPE concat.** Without head padding the W_UK bmm now writes into a `(B, N, L)` buffer through a transposed `out` view (verified: one GEMM kernel, no copy, identical result). For NoPE models (`qk_rope_head_dim == 0`) the FlashInfer sparse backend then uses `ql_nope` directly instead of `torch.cat` with a zero-width `q_pe`, which took the slow `CatArrayBatchedCopy` path (13.7 us/layer at 256 decode tokens, ~0.75 ms/layer per 16k-token prefill chunk).
3. **Glm5NextMoE: do not compute the router logits twice.** `MoERunner` holds the gate (passed via `FusedMoEFactory`) and recomputes the logits (`moe_runner.py:897`), so the gate GEMM in `Glm5NextMoE.forward` was wasted (two bf16->fp32 GEMMs + reductions per MoE layer). Follows `DeepseekV2MoE` and passes a placeholder.

## Test Plan

- Unit / kernel tests listed under "Tests" below.
- Serving benchmark, all builds back-to-back in one session on 4x GB300 (SM100), TP4, prefix caching disabled:
  ```bash
  vllm serve zai-org/GLM-5.3-Flash -tp 4 --attention-backend FLASHINFER_MLA_SPARSE \
    --max-model-len 69632 --max-num-seqs 256 --max-num-batched-tokens 16384 \
    --gpu-memory-utilization 0.9 --no-enable-prefix-caching
  vllm bench serve --backend vllm --model glm53-flash --dataset-name random --random-range-ratio 0 \
    --ignore-eos --seed 0 --random-input-len <in> --random-output-len <out> --max-concurrency <c> \
    --num-prompts <n> --num-warmups <c> --percentile-metrics ttft,tpot,itl,e2el
  ```
  Points: 1k/512 at c=1/64/256; prefill-only 8x2048, 2x8192, 8x32768, 2x65536; 32k/256 at c=16. The prefix-cache hit rate in the server log was checked to be 0.0% for every run.
- Accuracy per build: `lm_eval --model local-completions --tasks gsm8k --num_fewshot 5 --gen_kwargs temperature=0`

### Performance of this PR alone

main `156050598` vs main + this PR, 4x GB300, TP4, `--attention-backend FLASHINFER_MLA_SPARSE --max-model-len 69632 --max-num-seqs 256 --max-num-batched-tokens 16384`, **prefix caching disabled** (hit rate 0.0% checked in the server log), `vllm bench serve` random dataset with warmups, both servers run back-to-back in the same session. Decode tok/s is the steady-state window value (full concurrency); TPOT is the per-request median.

| point | metric | main | +#55736 | delta |
|---|---|---|---|---|
| 1k/512 c=1 | decode tok/s | 144 | 146 | +1.8% |
| 1k/512 c=1 | median TPOT ms | 6.57 | 6.46 | -1.7% |
| 1k/512 c=64 | decode tok/s | 2964 | 3090 | +4.3% |
| 1k/512 c=64 | median TPOT ms | 20.02 | 19.26 | -3.8% |
| 1k/512 c=256 | decode tok/s | 7147 | 7367 | +3.1% |
| 1k/512 c=256 | median TPOT ms | 33.87 | 32.87 | -3.0% |
| prefill 8x2048 | mean TTFT ms | 461 | 449 | -2.5% |
| prefill 2x8192 | mean TTFT ms | 389 | 380 | -2.4% |
| prefill 8x32768 | mean TTFT ms | 5258 | 5108 | -2.8% |
| prefill 2x65536 | mean TTFT ms | 2738 | 2661 | -2.8% |
| 32k/256 c=16 | decode tok/s | 287 | 296 | +3.1% |
| 32k/256 c=16 | mean TTFT ms | 3883 | 3835 | -1.2% |
| 32k/256 c=16 | median TPOT ms | 41.34 | 40.10 | -3.0% |

Decode: +3-4% tok/s (−3-4% TPOT) at c=64, c=256 and in the 32k-context point, −1.7% TPOT at c=1 (host-bound). Prefill: a consistent −2.5-2.8% TTFT on all four points

### Accuracy (per build, same session)

Each rung of the ablation was also evaluated on its own: **gsm8k** (1319 questions, 5-shot, greedy, lm_eval `local-completions`)

| build | gsm8k flexible-extract % | gsm8k strict-match % |
|---|---|---|
| main | 93.33 ± 0.69 | 93.10 ± 0.70 | 
| +#55736 | 93.03 ± 0.70 | 92.95 ± 0.71 |

## Duplicate-work check

`gh pr list --repo vllm-project/vllm --state open --search "fused_recurrent_kda stride"`, `"MoERunner gate router_logits Glm5Next"`, `"GLM-5.3-Flash prefill"`: no open PR covers these (the open GLM-5.3-Flash PRs #54951 / #55222 / #55385 / #55543 touch the indexer prefill sharding, SM90 sparse-MLA dtypes and the SM90 FA/FlashMLA wiring).

## Tests

- `pytest tests/models/glm5next/test_kda_recurrent.py` (new; fp32 PyTorch reference for plain and spec-decode shapes, strided-vs-contiguous bit-identity, unaddressable layouts rejected): 6 passed.
- bmm into a transposed `out`: max diff 0.0 vs `bmm().transpose()`, profiler shows a single `nvjet` kernel.
- `pre-commit run ruff-check / ruff-format` on the changed files: passed.
- e2e: serving smoke (chat sanity); prompt-logprob A/B on real 4k/12k/30k/60k prompts vs main is within ~1-2.5x the main-vs-main run-to-run noise floor with identical next tokens; **gsm8k 1319 questions (5-shot, greedy): main 93.03% ± 0.70, full series 92.65% ± 0.72** (5 questions, < 1σ; measured on the three PRs together).

## Notes for review

- AI assistance: this change was developed with Claude and reviewed by the submitter;

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: tests/kernels/test_glm5next_kda_recurrent_strided.py
Comment (gau-nernst) [inline tests/kernels/test_glm5next_kda_recurrent_strided.py:1]: I think we should stash this under `tests/models/glm5next`? similar to the recent models

File: tests/kernels/test_glm5next_kda_recurrent_strided.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+pytestmark = pytest.mark.skipif(
+    not current_platform.is_cuda(), reason="CUDA-only Triton kernel"
+)
+
+
+@pytest.mark.parametrize("num_seqs", [1, 7])
+@pytest.mark.parametrize("num_heads", [16])
+@pytest.mark.parametrize("head_dim", [128])
+def test_fused_recurrent_kda_strided_inputs_match_contiguous(
```
Comment (gau-nernst) [inline tests/kernels/test_glm5next_kda_recurrent_strided.py]: It's probably better to test against a reference implementation (e.g. pure PyTorch) instead of testing against itself

File: vllm/models/glm5next/nvidia/ops/third_party/kda/fused_recurrent.py [hunk 2]
```diff
@@ -265,6 +272,10 @@ def fused_recurrent_gated_delta_rule_fwd(
         stride_final_state_token=stride_final_state_token,
         stride_indices_seq=stride_indices_seq,
         stride_indices_tok=stride_indices_tok,
+        stride_q_t=H * K,
+        stride_k_t=H * K,
+        stride_v_t=HV * V,
+        stride_beta_t=HV * (V if beta.ndim == v.ndim else 1),
```
Comment (gau-nernst) [inline vllm/models/glm5next/nvidia/ops/third_party/kda/fused_recurrent.py]: Since you are adding strided support, might as well just pass `q.stride(something)` here?

File: vllm/models/glm5next/nvidia/ops/third_party/kda/kernels.py [hunk 3]
```diff
@@ -30,6 +30,24 @@
...
+    copy. Pure stride arithmetic: this runs eagerly per layer per step (the
+    KDA forward is a CUDA-graph break), so it must not create views.
+    """
+    st = x.stride()
+    if x.dim() == 4:
+        return st[3] == 1 and st[2] == x.shape[3]
+    return x.dim() == 3 and st[2] == 1
+
+
+def _token_stride(x: torch.Tensor, inner: int) -> int:
```
Comment (gau-nernst) [inline vllm/models/glm5next/nvidia/ops/third_party/kda/kernels.py]: `inner` doesn't do anything? Btw, I think you can do something like `assert x[0].is_contiguous()` as well to be more concise

File: vllm/models/glm5next/nvidia/ops/third_party/kda/kernels.py [hunk 4]
```diff
@@ -165,12 +192,14 @@ def fused_recurrent_kda(
...
 
+    # The kernel walks q/k/v/beta with an explicit token stride, so column
+    # slices of the fused projection buffer are consumed in place (no copy).
     o, final_state = fused_recurrent_kda_fwd(
-        q=q.contiguous(),
-        k=k.contiguous(),
-        v=v.contiguous(),
+        q=q if _token_strided(q) else q.contiguous(),
+        k=k if _token_strided(k) else k.contiguous(),
+        v=v if _token_strided(v) else v.contiguous(),
```
Comment (gau-nernst) [inline vllm/models/glm5next/nvidia/ops/third_party/kda/kernels.py]: (non-blocking) I think if you are already tackling layout issue, it might be better to assert certain contiguity instead of falling back to `.contiguous()` silently -> loud failure, so you know it's working / something else needs fixing.

File: tests/kernels/test_glm5next_kda_recurrent_strided.py (hunk 1 above)
Comment (JaredforReal) [reply tests/kernels/test_glm5next_kda_recurrent_strided.py]: done

**Comment by JaredforReal** [comment]
@gau-nernst PTAL, thanks!

File: tests/kernels/test_glm5next_kda_recurrent_strided.py
Comment (JaredforReal) [reply tests/kernels/test_glm5next_kda_recurrent_strided.py:1]: great! create a dir and buildkite test against glm5next

**Review by gau-nernst — APPROVED** [review]
LGTM

**Review by ZJY0516 — APPROVED** [review]
